### ISSUE TITLE
Serialize trait is unsound

### ISSUE THREAD

**Issue by dtolnay** [simd-lite/simd-json-derive#3]
This program demonstrates undefined behavior in safe code due to unsound code in the Serialize trait's default methods.

```rust
use simd_json_derive::Serialize;
use std::io::{self, Write};

struct Struct;

impl Serialize for Struct {
    fn json_write<W: Write>(&self, writer: &mut W) -> io::Result<()> {
        writer.write_all(b"\xff")
    }
}

fn main() {
    let _ = Struct.json_string().unwrap(); // violate utf8 invariant of String
}
```

**Comment by Licenser** [comment]
good catch! :D

**Comment by Licenser** [comment]
Fixed and published as 0.1.8, thanks for reporting this :D
